### PR TITLE
Phase 4 - Simple Java application that can add a bunch of posts in your local MySQL

### PR DESCRIPTION
**Pull request by kanghowoo**
Let’s create another simple Java application that can add a bunch of posts in your local MySQL.
When you run this Java application with count N parameter, you can insert a bunch of posts in your local MySQL.
for example)
```
java -jar … BulkInsertApp.jar --count 100000
```
With this application, insert 300,000 posts in local PC.

### CODE REVIEW

File: src/main/java/BulkInsertApp.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    private int count;
+
+    public static void main(String[] args) {
+        BulkInsertApp app = new BulkInsertApp();
+        JCommander.newBuilder().addObject(app).build().parse(args);
+        app.run();
+    }
+
+    private void run() {
+        String insertSQL = "INSERT INTO boards"  + " (title, content, user_id) VALUES (?, ?, ?)";
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java]: Let's exit with code 1 when count is smaller than 1

File: src/main/java/BulkInsertApp.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        final String password = String.valueOf(dataSource.get("password"));
+
+        try (Connection connection = DriverManager.getConnection(url, user, password)) {
+            connection.setAutoCommit(false);
+            try (PreparedStatement preparedStatement = connection.prepareStatement(insertSQL)) {
+                for (int i = 1; i <= count; i++) {
+                    preparedStatement.setString(1, "test_title" + i);
+                    preparedStatement.setString(2, "test_content" + i);
+                    preparedStatement.setInt(3, USER_ID);
+                    preparedStatement.addBatch();
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java]: Why did you not use INSERT bulk query? I think bulk insert is more efficient.
Please check what is the difference between bulk INSERT query and N INSERT queries with batch.
And fix this to use a single bulk INSERT query per 1000 items.

File: src/main/java/BulkInsertApp.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    private void run() {
+        String insertSQL = "INSERT INTO boards"  + " (title, content, user_id) VALUES (?, ?, ?)";
+
+        Map<String, Object> config = yamlConfigReader();
+        Map<String, String> dataSource = (Map<String, String>) config.get("datasource");
+        final String url = dataSource.get("url");
+        final String user = dataSource.get("user");
+        final String password = String.valueOf(dataSource.get("password"));
+
+        try (Connection connection = DriverManager.getConnection(url, user, password)) {
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java]: What is this try-resource syntax? Can you explain me what's the difference with just try-catch ?

File: src/main/java/BulkInsertApp.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                    preparedStatement.setInt(3, USER_ID);
+                    preparedStatement.addBatch();
+
+                    if (i % 1000 == 0 || i == count) {
+                        preparedStatement.executeBatch();
+                        connection.commit();
+                    }
+                }
+            } catch (SQLException e) {
+                connection.rollback();
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java:70]: return the error code if it failed and at least log something.

File: src/main/java/BulkInsertApp.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java]: ditto

**Review by jinhyukify — COMMENTED** [review]
Okay nice. Left some comments

File: src/main/java/BulkInsertApp.java (hunk 2 above)
Comment (kanghowoo) [reply src/main/java/BulkInsertApp.java]: Bulk query faster than using separate single-row [INSERT] statements. I found it in the following mysql document. 
[https://dev.mysql.com/doc/refman/8.0/en/insert-optimization.html](url)

According to the document, sending one large statement will save almost 7 per insert statement.

File: src/main/java/BulkInsertApp.java (hunk 3 above)
Comment (kanghowoo) [reply src/main/java/BulkInsertApp.java]: `try-with-resources` can resources automatically closed. But, `try-catch` can't. 
So manually close resources with null checking.  
As a result, we can make the code more concise and prevent mistakes

Comment (jinhyukify) [reply src/main/java/BulkInsertApp.java]: Nice one James

File: src/main/java/BulkInsertApp.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            connection.setAutoCommit(false);
+
+            try (Statement statement = connection.createStatement()) {
+                for (int i = 1; i <= (count / BATCH_SIZE) + 1; i++) {
+                    StringBuilder sql = new StringBuilder("INSERT INTO boards (title, content, user_id) VALUES ");
+
+                    for (int j = 1; j <= BATCH_SIZE; j++) {
+                        sql.append("('")
+                           .append("test_title").append(i).append("', '")
+                           .append("test_content").append(i).append("', '")
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java]: Then title and content will be same for all rows in same BATCH. Is this what you intended?
1\~1000 rows -> test_title1, test_content1
1000\~2000 rows -> test_title2, test_content2

File: src/main/java/BulkInsertApp.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                           .append(USER_ID)
+                           .append("')");
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java]: No need to fix to this.
나같으면 그냥 간단하게 이렇게 짰을듯. 가독성이 조금안좋아서 (만약 위 코드와 같다면)
```suggestion
                   final String title = "test_title" + i;
                   final String content = "test_content" + i;
                    for (int j = 1; j <= BATCH_SIZE; j++) {
                        sql.append(String.format("('%s', '%s', %d)", title, content, USER_ID));
```

File: src/main/java/BulkInsertApp.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Map;
+
+import org.yaml.snakeyaml.Yaml;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.Parameter;
+
+public class BulkInsertApp {
+
+    public static final int USER_ID = 1;
```
Comment (jinhyukify) [inline src/main/java/BulkInsertApp.java]: 뭐 지금은 1이라 동작에 상관없긴한데 `user_id` column type이 8 byte 짜리였나? 그래도 타입은 맞추는게 좋겠는데..

**Comment by jinhyukify** [comment]
Let's add dryRun mode.
In LINE, we always add **dryRun** mode to this kind of one-time-run-application.
And before running the application, we usually run with `dryRun` mode first and check this will work well or not first.
(For example)
```
java -jar … BulkInsertApp.jar --count 100000 --dry-run false
java -jar … BulkInsertApp.jar --count 100000 --dry-run true
```

Let's add `--dry-run` parameter and this should be default to *true**
When dryRun mode is active(=true), then this program **should not** actually query to database, but just print the query to standard output or to file.
When dryRun mode is false, then your application will work like you developed now. (query to database and insert data)

File: src/main/java/BulkInsertApp.java (hunk 6 above)
Comment (kanghowoo) [reply src/main/java/BulkInsertApp.java]: That's my mistake. I fixed it.

**Review by jinhyukify — APPROVED** [review]
LGTM let's go to next phase

**Comment by jinhyukify** [comment]
Please merge this

**Pull request merged**